frontend/viewer/src/lib/writing-system-service.ts

```diff
@@ -0,0 +1,135 @@
+  ): IWritingSystem[] {
+    ws = ws ?? 'vernacular-analysis';
+    switch (ws) {
+      case 'vernacular-analysis':
+        return [...new Set([...this.writingSystems.vernacular, ...this.writingSystems.analysis].sort())];
+      case 'analysis-vernacular':
+        return [...new Set([...this.writingSystems.analysis, ...this.writingSystems.vernacular].sort())];
+      case 'first-analysis':
+        return [this.writingSystems.analysis[0]];
+      case 'first-vernacular':
+        return [this.writingSystems.vernacular[0]];
+      case 'vernacular':
+        return this.writingSystems.vernacular;
+      case 'analysis':
+        return this.writingSystems.analysis;
+    }
+    console.error(`Unknown writing system selection: ${ws as string}`);
+    return [];
+  }
+
+  indexExemplars(): string[] | undefined {
+    return this.defaultVernacular()?.exemplars;
+  }
+
+  headword(entry: IEntry, ws?: string): string {
+    if (ws) {
+      return WritingSystemService.headword(entry, ws) || '';
+    }
+
+    return firstTruthy(this.vernacular, ws => WritingSystemService.headword(entry, ws.wsId)) || '';
+  }
+
+  private static headword(entry: IEntry, ws: string): string | undefined {
+    return entry.citationForm[ws] || entry.lexemeForm[ws];
+  }
+
+  pickBestAlternative(value: IMultiString, wss: 'vernacular' | 'analysis'): string
+  pickBestAlternative(value: IMultiString, firstChoice: IWritingSystem): string
+  pickBestAlternative(value: IMultiString, firstChoice: IWritingSystem | 'vernacular' | 'analysis'): string {
+    let allWs: IWritingSystem[];
+    if (typeof firstChoice === 'object') {
+      allWs = [firstChoice, ...this.allWritingSystems()];
+    } else {
+      switch (firstChoice) {
+        case 'vernacular':
+          allWs = this.allWritingSystems('vernacular-analysis');
+          break;
+        case 'analysis':
+          allWs = this.allWritingSystems('analysis-vernacular');
+          break;
+        default:
+          throw new Error(`Unknown writing system selection ${firstChoice as unknown as string}`);
+      }
+    }
+
+    return this.first(value, allWs) || '';
+  }
+
+  getWritingSystem(wsId: string, selection?: 'vernacular' | 'analysis'): IWritingSystem | undefined {
+    const writingSystems = this.pickWritingSystems(selection);
+    return writingSystems.find(ws => ws.wsId === wsId);
+  }
+
+  firstGloss(sense: ISense): string {
+    return this.first(sense.gloss, this.analysis) || '';
+  }
+
+  firstDef(sense: ISense): string {
+    return this.first(sense.definition, this.analysis) || '';
+  }
+
+  glosses(entry: IEntry | undefined): string {
+    if (!entry?.senses?.length) return '';
+    return entry.senses.map(sense => this.first(sense.gloss, this.analysis)).filter(gloss => !!gloss).join(', ');
+  }
+
+  firstDefOrGlossVal(sense: ISense | undefined): string {
+    if (!sense) return '';
+    return this.first(sense.definition, this.analysis) || this.first(sense.gloss, this.analysis) || '';
```

Comment:
not sure the best answer, but this is doing a depth first check, by that I mean it's going to pull anything from definition before it pulls from gloss, but it might make more sense (depending on context) to pull from the default ws first. If we were to rewrite this with iteration, do we iterate the writing systems in the outer loop or the properties? I'm not sure but I don't want to ignore one for the other. This wasn't possible in the previous version of this function since there wasn't an order to the ws being checked.
This also applies to the example sentence version below.

Reply:
I did think about this briefly as I refactored.
I'm pretty sure there are places where we should prefer glosses or prefer the default writing system vs a definition.

Can we revisit this?
This is called in various places. Some of them probably shouldn't be calling it at all.
I'm not sure what each of them really wants.